File: src/RSSQuick/Models/FeedItem.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace RSSReaderWPF
{
    /// <summary>
    /// A node in the feed tree: either a feed, or a folder holding more of them.
    /// </summary>
    public class FeedItem : INotifyPropertyChanged
    {
        private string _title = string.Empty;
        private string _url = string.Empty;
        private string _category = string.Empty;
        private bool _isCategory;

        public string Title
        {
            get => _title;
            set { _title = value; OnPropertyChanged(); }
        }

        /// <summary>The feed URL. Empty for a folder.</summary>
        public string Url
        {
            get => _url;
            set { _url = value; OnPropertyChanged(); }
        }

        public string Category
        {
            get => _category;
            set { _category = value; OnPropertyChanged(); }
        }

        /// <summary>True for a folder, which loads every feed beneath it rather than one.</summary>
        public bool IsCategory
        {
            get => _isCategory;
            set { _isCategory = value; OnPropertyChanged(); }
        }

        public ObservableCollection<FeedItem> Children { get; } = new();

        public event PropertyChangedEventHandler? PropertyChanged;

        protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

File: macos/Sources/RSSQuickCore/FeedItem.swift
import Foundation

/// A node in the feed tree: either a feed, or a folder holding more of them.
///
/// A reference type because `NSOutlineView` identifies its rows by object identity - it asks for
/// `child(_:ofItem:)` and then hands the same item back in every later call. It is immutable, so
/// it is safely `Sendable` and can be handed to a background load without copying.
public final class FeedItem: Sendable {
    public let title: String

    /// The feed URL. Empty for a folder.
    public let url: String

    public let category: String

    /// True for a folder, which loads every feed beneath it rather than one.
    public let isCategory: Bool

    public let children: [FeedItem]

    public init(
        title: String,
        url: String = "",
        category: String = "",
        isCategory: Bool = false,
        children: [FeedItem] = []
    ) {
        self.title = title
        self.url = url
        self.category = category
        self.isCategory = isCategory
        self.children = children
    }

    /// Every feed under this folder, at any depth. A feed answers with itself.
    public var allFeeds: [FeedItem] {
        guard isCategory else { return [self] }
        return children.flatMap(\.allFeeds)
    }
}
